### PR TITLE
Make `bias` a ReplicatedTensor in sharded linear.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #77307

`bias` was implicitly considered a ReplicatedTensor as part of the
sharded linear implementation.

Since, we now have ReplicatedTensor to express this and also support for DDP,
we can express this as ReplicatedTensor and have our ShardedTensor +
ReplicatedTensor inter-op deal with the associated ops in sharded linear.

Differential Revision: [D36329170](https://our.internmc.facebook.com/intern/diff/D36329170/)

### CODE REVIEW

File: test/distributed/_shard/sharded_tensor/ops/test_linear.py [hunk 1]
```diff
@@ -125,22 +136,26 @@ def _run_sharded_linear(
         optim.step()
         self.assertNotEqual(previous, local_linear.weight)
         previous_sharded_weight = sharded_weight.clone()
-        previous_sharded_bias = sharded_linear.bias.clone()
+        previous_sharded_bias = sharded_linear.module.bias.clone()
+
+        # Use different learning rate (scale by world_size) for bias since it is replicated via DDP.
+        bias_optim = torch.optim.SGD([sharded_linear.module.bias], lr=0.1 * self.world_size)
```
Comment (wanchaol) [inline test/distributed/_shard/sharded_tensor/ops/test_linear.py:142]: why we need to use different learning rate instead of a single one? Im not quite sure I can get it, is this also what the end user code look like?

File: torch/distributed/_shard/partial_tensor.py [hunk 2]
```diff
@@ -233,6 +234,27 @@ def _transpose_impl(types, args=(), kwargs=None):
...
+def partial_add(types, args=(), kwargs=None):
+    if isinstance(args[0], _PartialTensor) and isinstance(args[1], ReplicatedTensor):
+        world_size = dist.get_world_size(args[0].process_group)
+        with torch._C.DisableTorchFunction():
+            lhs = args[0].local_shard
+            rhs = args[1]
+            if lhs.dim() == 1 and rhs.dim() == 1:
+                if lhs.numel() % rhs.numel() != 0:
+                    raise RuntimeError(
+                        f"torch function '{func.__name__}', with args: {args} and "
```
Comment (wanchaol) [inline torch/distributed/_shard/partial_tensor.py:247]: nit: fix lints

File: test/distributed/_shard/sharded_tensor/ops/test_linear.py (hunk 1 above)
Comment (pritamdamania87) [reply test/distributed/_shard/sharded_tensor/ops/test_linear.py:142]: This is covered in our DDP docs: https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel

```
When a model is trained on M nodes with batch=N, the gradient will be M times smaller when compared to the same model trained on a single node with batch=M*N if the loss is summed (NOT averaged as usual) across instances in a batch (because the gradients between different nodes are averaged). You should take this into consideration when you want to obtain a mathematically equivalent training process compared to the local training counterpart. But in most cases, you can just treat a DistributedDataParallel wrapped model, a DataParallel wrapped model and an ordinary model on a single GPU as the same (E.g. using the same learning rate for equivalent batch size).
```

Comment (wanchaol) [reply test/distributed/_shard/sharded_tensor/ops/test_linear.py:142]: I see, does it mean that we are only doing this (initializing two optimizers) purely for local megatron numerical parity, but for most cases, we don't need to do it with DDP?

**Review by wanchaol — APPROVED** [review]
looks good!

File: test/distributed/_shard/sharded_tensor/ops/test_linear.py (hunk 1 above)
Comment (fduwjj) [reply test/distributed/_shard/sharded_tensor/ops/test_linear.py:142]: Numerically, they are the same. But isn't this going to confuse users because we need to use two learning rates for bias and weight. I am not sure if using separate learning rate is a common practice or not.

File: test/distributed/_shard/sharded_tensor/ops/test_linear.py [hunk 3]
```diff
@@ -65,29 +67,24 @@ def _run_sharded_linear(
         # Shard the parameter.
         shard_parameter(sharded_linear, "weight", spec)
 
+        # Wrap with DDP enabled with ReplicatedTensor
+        with _ddp_replicated_tensor(True):
+            sharded_linear = DDP(sharded_linear)
```
Comment (fduwjj) [inline test/distributed/_shard/sharded_tensor/ops/test_linear.py:72]: Does this DDP will automatically wrap `bias` as a Replicated Tensor? What if users don't use DDP, does that mean we don't even allow users to use `bias` as a `torch.Tensor`?

File: test/distributed/_shard/sharded_tensor/ops/test_linear.py (hunk 1 above)
Comment (pritamdamania87) [reply test/distributed/_shard/sharded_tensor/ops/test_linear.py:142]: I think we will probably have to document this similar to how DDP above documents how to use an appropriate learning rate.

File: test/distributed/_shard/sharded_tensor/ops/test_linear.py (hunk 3 above)
Comment (pritamdamania87) [reply test/distributed/_shard/sharded_tensor/ops/test_linear.py:72]: It doesn't do it automatically, you need `with _ddp_replicated_tensor(True):`. The tricky thing is that if users don't use DDP, we kinda end up assuming `bias` is still replicated.

I do feel it makes things cleaner if users always use DDP and we recommend users to always combine DDP with sharding and not use sharding alone, because then it results in things like us treating `bias` implicitly as a ReplicatedTensor and I feel its better to have things be explicit.

File: test/distributed/_shard/sharded_tensor/ops/test_linear.py (hunk 1 above)
Comment (fduwjj) [reply test/distributed/_shard/sharded_tensor/ops/test_linear.py:142]: I guess what I want to say here is that, this might still cause confusions to users even if we document it. Hopefully, if we can explain it well in the document so that users won't have too many questions.
